File: evo.rb
#all characters available for building our strings
CHARS = [ *('a' .. 'z'), *('A' .. 'Z'), ' ']

def random_char
	return CHARS[rand(CHARS.size-1)]
end

#return random variation for mutation
def variation var
	rand(-var .. var)
end

#randomly change one character in the string
def mutate str, var=1
	mutated = str.dup
	pos = rand(mutated.size)
	ch = mutated[pos]
	#mutate the charater at pos
	mutated[pos] = (ch.ord + variation(var) ).chr
	return mutated
end

#use only characters from the target string, for mutation
def mutate4target str, target
	mutated = str.dup
	target_chars = target.split('').uniq
	pos = rand(mutated.size)
	ch = mutated[pos]
	mutated[pos] = target_chars[rand(target_chars.size)]
	return mutated
end

def fitness str1, str2
	dist = 0
	for i in 0 .. str1.size-1
		dist += (str1[i].ord - str2[i].ord) ** 2
	end
	return dist
end

#crossover two strings
def mate str1, str2
	new_str = str1.dup
	start = rand(str1.size-1)
	stop  = rand(str1.size-1)
	start,stop = stop,start	if start > stop
	new_str[start,stop] = str2[start,stop]
	return new_str
end

class Evo
	attr_accessor :target, :pool, :found_at, :mutation, :mutation_variation

	def initialize target, a={}
		@target = target
		@pool = []
		@display_every = a[:display_every] || 10
		@iterations = a[:iterations] || 100
		@found_at = a[:found_at] || 0
		@mutation = a[:mutation] || :basic
		@mutation_variation = a[:mutation_variation] || 1
	end

	def add2pool str
		f = fitness @target, str
		@pool.push Hash[ :data => str, :fitness => f ]
	end

	def gen_pool len
		@pool = []
		for i in 1 .. len
			str = ''
			for _ in 1 .. @target.size ; str += random_char end
			self.add2pool str
		end
	end

	def display_pool
		puts "pool> " + (@pool.map {|e| e[:data]}   ).join(",")
#		puts "fit > " + (@pool.map {|e| e[:fitness]}).join(",")
	end

	#pick parent with tendency of smaller fitness i.e. close to target
	def rand_parent range; ((rand * rand) * range).to_i end
	def pick_parents
		range = @pool.size #-1
		pidx1 = rand_parent range
		pidx2 = rand_parent range
		#just quick precaution so that we dont pick the same parent twice
		pidx2 = pidx1 == pidx2 ? rand_parent(range) : pidx2

		return pidx1,pidx2
	end

	def evolve
		#insure that pool is sorted
		@pool.sort_by! { |e| -e[:fitness] } #reverse order by fitness
		parent1, parent2 = pick_parents
		#did we find match
		return true if @pool[0][:fitness] == 0
		#the pool is sorted, worse -to-> good
		child = mate @pool[parent1][:data], @pool[parent2][:data]
		#pick different mutation methods
		if @mutation == :basic
			mutated = mutate child
		else
			mutated = mutate4target child, @target
		end
		#puts "ev> " + @pool[parent1][:data] + " : " + @pool[parent2][:data] + ' = ' + mutated
		fit_score = fitness @target, mutated
		#die off, replace the worse parent with the child (if better)
		if @pool[0][:fitness] > fit_score
			@pool[0] = Hash[:data => mutated, :fitness => fit_score ]
		end

		return false
	end

	def iterate
		raise "Please generate gene pool first" if @pool.size == 0
		for i in 1 .. @iterations
			#puts ">#{i}"
			if evolve
				@found_at = i
				return i
			end
			display_pool if i % @display_every == 0
		end
		return false
	end

end

e = Evo.new(target = 'hello', iterations: 9000, display_every: 100, mutation_variation: 1)
e.gen_pool 20
e.display_pool

at = e.iterate
if at
	puts "evolved match #{at}: #{e.pool[0]}"
else
	puts "unable to evolve"
end
puts "mutation method : #{e.mutation} , #{e.mutation_variation}"

File: evo.pl
#!/usr/bin/env perl
use strict;
use v5.10;
use Data::Dumper;

package Evo;

#used to pick random chars
our @CHARS = ('A' .. 'Z', 'a' .. 'z', ' ');

sub new {
	my $class = shift;
	my $self = { };
	bless $self, $class;
	$self->init(@_);
	return $self;
}

sub init {
	my ($self, %a) = @_;
	$$self{target} = $a{target};
	$$self{pool} = $a{pool} || [];
	$$self{display_every} = $a{display_every} || 10;
	$$self{steps} = $a{steps} || 1500;
}

sub random_char { $CHARS[rand $#CHARS] }

sub mutate {
	my ($str,$range) = @_;
	$range ||= 1;
	my $len = length $str;
	my $pos = int(rand $len);#pick random pos
	my @range = -$range .. $range;
	my $direction = $range[ rand( (2 * $range) + 1) ];
	my $char = substr($str, $pos, 1);
	my $ord = ord($char) + $direction;
	my $new_char = chr($ord);
	#replace a character
	substr($str, $pos, 1) = $new_char;
	return $str #mutated string
}

sub fitness {
	my ($str1,$str2) = @_;
	my $len = length $str1;
	my $val = 0;
	for my $i (0 .. $len-1) {
		#calculate eucledian distance
		$val += ( ord(substr $str1, $i, 1) - ord(substr $str2, $i, 1) ) ** 2;
	}
	return $val;
}

sub mate {
	my ($str1,$str2) = @_;
	my $new_str = $str1;
	my $len = length $str1;
	my $start = int(rand($len));
	my $stop  = int(rand($len));
	($start,$stop) = ($stop,$start) if $start > $stop;
	my $width = $stop - $start;
	substr $new_str, $start, $width, ( substr $str2,$start,$width );
	return $new_str
}

#============ Class methods ===========================================


sub display_pool {
	my $self = shift;
	say "pool>", join ', ', map { $$_{data} } @{$$self{pool}};
}

sub add2pool {
	my ($self, $str) = @_;
	my $fitness = 999_999_999;
	if ($$self{target}) {
		$fitness = fitness $str, $$self{target}
	}
	push @{$$self{pool}}, { data => $str, fitness => $fitness }
}


#generate random pool
sub gen_pool {
	my ($self,$size,$len) = @_;
	$len ||= length $$self{target};
	for my $el ( 1 .. $size ) {
		my $str = '';
		$str .= random_char for 1 .. $len;
		$self->add2pool($str);
	}
}

sub rand_pick { int( (rand * rand) * $_[0]) }
#select two parents for mixing, the lower the score better fitted for selection
sub pick_parents {
	my $self = shift;
	#two random nums with tendence to get higher numbers (!fixme make the random values unique, non-repeating)
	my $size = scalar @{$$self{pool}};
	my $rand1 = rand_pick($size);
	my $rand2 = rand_pick($size);
	$rand2 =  $rand1 == $rand2 ? rand_pick($size) : $rand2;

	#pick two with preference for lower dist (keys are dist, values are pool idx)
	@{$$self{sorted_idxs}} = sort { $$self{pool}[$b]{fitness} <=> $$self{pool}[$a]{fitness} } 0 .. $#{$$self{pool}};#reverse order
	my ($pidx1,$pidx2) = @{$$self{sorted_idxs}}[ $rand1, $rand2 ];
	#say "parents> $pidx1:$f1, $pidx2:$f2";
	return $pidx1, $pidx2
}


sub selection {
	my ($self,$mutated) = @_;
	#pick who will die, prefably the one furthest away
	my $die_idx = $$self{sorted_idxs}[0]; #max distance idx
	my $fit_score = fitness $$self{target}, $mutated;
	#replace parent with child in the population, but only if better genes
	if ($$self{pool}[$die_idx]{fitness} > $fit_score ) {
		$$self{pool}[$die_idx] = {data => $mutated, fitness => $fit_score }
	}
}

sub evolve {
	my ($self, $idx) = @_;

	#found it
	return 1 if $$self{pool}[ $$self{sorted_idx}[-1] ]{fitness} == 0;

	### CROSSOVER ###
	my ($pidx1,$pidx2) = $self->pick_parents();#find candidates for sex
	my $child = mate $$self{pool}[$pidx1]{data}, $$self{pool}[$pidx2]{data};

	### MUTATION ###
	my $mutated = mutate $child; #now introduce a mutation
	#say "ev> " . $$self{pool}[$pidx1]{data} . ' : ' . $$self{pool}[$pidx2]{data} . ' = ' . $mutated;

	### SELECTION ###
	$self->selection($mutated);

	return 0 #have not reached the target
}

sub iterate {
	my $self = shift;
	for my $iter (0 .. $$self{steps}) {
		#say "--> $iter";
		if ( $self->evolve ) {
			$$self{iterations} = $iter;
			return 1;
		}
		$self->display_pool if $iter % $$self{display_every} == 0;
	}
	$$self{iterations} = $$self{steps};
	return 0
}



package main;

my $e = new Evo(target => 'hello world', display_every => 100, steps => 5000);
say;
$e->gen_pool(10);
$e->display_pool;

say ">>>>>>>>>>>>>>>>>>>>>>";

if ( $e->iterate ) {
	$e->display_pool;
	say "evolved idx> " . Dumper( $$e{pool}[0] );
	say "iterations : ", $$e{iterations};
} else {
	say "could not evolve in that many iterations : ", $$e{iterations}
}
